system-addon/lib/UserDomainAffinityProvider.jsm

```diff
@@ -0,0 +1,262 @@
+        // => "{a.com":{"ps1":2,"ps2":2}, "b.com":{"ps1":3,"ps2":3}}""
+        .reduce((acc, cur) => this._combineScores(acc, cur));
+  }
+
+  /**
+   * Calculates the user domain affinity scores for the given time segment.
+   *
+   * @param ts The time segment
+   * @return The parameter specific scores for all domains with visits in
+   * this time segment: {"a.com": {"ps1": 12, "ps2": 34}, "b.com" ...}
+   */
+  calculateUserDomainAffinityScores(ts) {
+    // Returns domains and visit counts for this time segment: {"a.com": 1, "b.com": 2}
+    let visits = this.queryVisits(ts);
+
+    return Object
+      .keys(visits)
+      .reduce((acc, d) => merge(acc, {[d]: this.calculateScoresForParameterSets(ts, visits[d])}), {});
+  }
+
+  /**
+   * Calculates the scores for all parameter sets for the given time segment
+   * and domain visit count.
+   *
+   * @param ts The time segment
+   * @param vc The domain visit count in the given time segment
+   * @return The parameter specific scores for the visit count in
+   * this time segment: {"ps1": 12, "ps2": 34}
+   */
+  calculateScoresForParameterSets(ts, vc) {
+    return Object
+      .keys(this.parameterSets)
+      .reduce((acc, ps) => merge(acc, {[ps]: this.calculateScoreForParameterSet(ts, vc, this.parameterSets[ps])}), {});
+  }
+
+  /**
+   * Calculates the final affinity score in the given time segment for the given parameter set
+   *
+   * @param timeSegment The time segment
+   * @param visitCount The domain visit count in the given time segment
+   * @param parameterSet The parameter set to use for scoring
+   * @return The final score
+   */
+  calculateScoreForParameterSet(timeSegment, visitCount, parameterSet) {
+    return this.calculateScore(
+      visitCount * parameterSet.timeSegmentWeights[timeSegment.id],
+      parameterSet.perfectFrequencyVisits,
+      parameterSet.frequencyFactor);
+  }
+
+  _combineScores(a, b) {
+    // Merge both score objects so we get a combined object holding all domains.
+    // This is so we can combine them without missing domains that are in a and not in b and vice versa.
+    const c = merge({}, a, b);
+    return Object.keys(c).reduce((acc, d) => merge(acc, this._combine(a, b, c, d)), {});
+  }
+
+  _combine(a, b, c, d) {
+    return Object
+      .keys(c[d])
+      // Summing up the parameter set specific scores of each domain
+      .map(ps => ({[d]: {[ps]: Math.min(1, ((a[d] && a[d][ps]) || 0) + ((b[d] && b[d][ps]) || 0))}}))
+      // Reducing from an array of objects with a single parameter set to a single object
+      // [{"a.com":{"ps1":11}}, {"a.com: {"ps2":12}}] => {"a.com":{"ps1":11,"ps2":12}}
+      .reduce((acc, cur) => ({[d]: merge(acc[d], cur[d])}));
+  }
+
+  /**
+   * Calculates a value on the curve described by the provided parameters. The curve we're using is
+   * (a^(bx) - 1) / (a^b -1): https://www.desmos.com/calculator/maqhpttupp
```

Comment:
nit: I had some difficulty to associate the variables in this formula with the following function arguments. For example, I couldn't find `b` in the function, and this is no `y` in this formula. Could you clarify or reorganize it please?

Reply:
Thanks, clarified and improved jsdocs.